Customize buttons

arooooon opened the pull request:
Feedback is welcome. 


Todo/problems:
* Backwards compatibility for reviewer settings.
* Submenu items don't care about showAsAction.
* More actions.
* Icons
* Actionbar/overflow doesn't update on device rotation.

https://github.com/ankidroid/Anki-Android/issues/3154

timrae commented:
@arooooon sorry for the delay I'm quite busy with job interviewing at the moment. The code looks like a good first step, though I haven't had a chance to test it properly yet. Do you want to keep adding to this PR or merge (after testing) and make new ones? Is there anything in specific that you wanted feedback on?

arooooon commented:
Hey I understand, I have stuff to do too :). I am glad the code seems ok so far. I will continue commiting to this PR and then squash em when it is good enough.

eginhard commented:
Oh, I really like this and it works great already!

timrae commented:
@eginhard Do you think we should merge like this? Please proceed with your own judgement, the only reason I haven't merged is due to not testing it yet :)

timrae commented on AnkiDroid/src/main/res/values/10-preferences.xml:
```diff
@@ -140,6 +140,9 @@
     <string name="enable_api_title">Enable AnkiDroid API</string>
     <string name="enable_api_summary">Let other apps connect to AnkiDroid and read / write data without your intervention</string>
+    <string name="custom_buttons">Customize buttons</string>
+    <string name="custom_buttons_summary">Customize which buttons appear in the reviewer action bar.</string>
```
According to the Android style guide, single sentences should not be terminated with a period

timrae commented on AnkiDroid/src/main/res/values/11-arrays.xml:
```diff
@@ -40,6 +40,12 @@
+        <item>Always show</item>
+        <item>If room</item>
+        <item>Menu only</item>
+        <item>Disabled</item>
```
Since we already have a string "Disabled", it would be better to reference it via @string. In general it's better to do this for all strings in arrays, as it makes translations robust against changes to the order of the items in the arrays

arooooon replied:
true, though there isn't a single "@string" reference in arrays right now ;)

timrae replied:
Ah yeah you have to do all or nothing, otherwise you end up with "@string/..." getting up for translation. Check out constants.xml which has lots of arrays with @string refs, as those don't get sent for translation

timrae commented on AnkiDroid/src/main/res/xml/preferences_custom_buttons.xml:
```diff
@@ -0,0 +1,93 @@
+~ this program.  If not, see <http://www.gnu.org/licenses/>.
+-->
+
+<!-- Navigation Prefrences -->
```
This is an old comment, and you don't need to reference everyone's names in the copyright (just your own).

arooooon replied:
Ah yes I based that whole file on preferences_gestures so that is why the copyright followed.

timrae commented:
These comments are kind of nitpicking so I hesitate to make them for a casual contributor, but it seems you are keen to participate in the project so if you don't mind fixing them up that would be appreciated :)

arooooon commented:
That's ok and yes it's kinda fun to contribute actually.

There is really no point in merging right now, I am still working on it and currently it is not even compatible on older APIs ( crash).

I still have no idea how to fix submenu items to act on showAsAction so I currently have made an ugly workaround.

timrae commented:
For the crash, I think it's easiest to just do it programmatically using the Compat method I built. See the [custom sync server pull request](https://github.com/ankidroid/Anki-Android/pull/4026/files#diff-37ecac5b8e8b7ef745b768cb74a625b8R286).

arooooon commented:
Ok new commit, still not done. Just fixed a few things.

timrae commented:
For the bury/delete/etc submenu items, how would you like this to ideally work if no workarounds were necessary? Is it necessary to give users the ability to make each of these submenu items available as standalone items from a single tap, or do you think it would be enough to just have one button in the action bar that shows a submenu?

arooooon commented:
Yes I absolutely think we should make bury etc.. single tapable. Though I don't think "ifRoom" is crucial.

I have looked into ActioProvider and I can't see how it can help ? Does it's submenu items behave differently or is it possible to see if an ActionProvider is in the actionbar or overflow?

Also I copied your compatibility solution, super!

Hope I didn't miss any fixes!

timrae commented:
> Yes I absolutely think we should make bury etc.. single tapable.

I agree in principle, but I feel there are currently a couple of issues with the implementation that we may be able to improve upon, and would like to discuss a little bit more.

My main concern is that I suspect many users (myself included) are just as likely to want to suspend or bury a note as they are a card. This isn't really feasible in the current implementation as they both have the same icon, and I think the action bar would become too cluttered anyway even if someone were to come up with clever designs for the icons that could convey the same concept as the text.

What do you think about the following suggestion?

Instead of having 5 new menu items, we just have 3: suspend, bury, and delete. When suspend or bury is pressed, we calculate `len(mNote.cards())` and if it's greater than 1 then we show a submenu that lets the user choose between suspending / burying only the current card, or the current card and all of its siblings as well. We could probably actually do a bit better and only show the submenu if the sibling cards are not already suspended / buried.

An alternative simpler idea could be to only give an icon for the "card" versions of these actions (I think these are the most intuitive as they only act on the current card that you're looking at), or maybe to keep it how it is, but also provide a single unified "hide / delete" item in the action bar which shows the submenu (reducing number of taps from 3 to 2 for any of these actions without cluttering the action bar).

One other thing I noticed is that show/hide whiteboard, and clear whiteboard are not included in the list. With all these new configuration options, I think users would probably require the ability to set either of these actions to "always". This would only take effect when the whiteboard is enabled obviously. Would you be able to add them please? You should also remove [this block of code](https://github.com/arooooon/Anki-Android/blob/customButtonsAGAIN/AnkiDroid/src/main/java/com/ichi2/anki/Reviewer.java#L272) from Reviewer.java that tries to do clever things based on the size of the screen. 

Other than those points, it's looking great. When you're done, could you please squash down to one commit using git rebase in interactive mode and do a force push? Then I'll go ahead and merge.

arooooon commented:
> Instead of having 5 new menu items, we just have 3: suspend, bury, and delete. When suspend or bury is pressed, we calculate len(mNote.cards()) and if it's greater than 1 then we show a submenu that lets the user choose between suspending / burying only the current card, or the current card and all of its siblings as well. We could probably actually do a bit better and only show the submenu if the sibling cards are not already suspended / buried.

This is what I decided to do.

> One other thing I noticed is that show/hide whiteboard, and clear whiteboard are not included in the list. With all these new configuration options, I think users would probably require the ability to set either of these actions to "always". This would only take effect when the whiteboard is enabled obviously. Would you be able to add them please? You should also remove this block of code from Reviewer.java that tries to do clever things based on the size of the screen (overriding the setting in the menu xml file). 

I added all the whiteboard stuff and removed some code. I don't know if that was exactly what you meant?


I also added ifRoom for Suspended, Bury and delete. Though if you do choose "if room" in settings, these items will not appear in the Hide/delete submenu when they are not in the action bar. By default they are in the submenu though.


Let me know what you think and I will squash it down.

Know issues:
* Activate whiteboard icon is the same as Hide/Show whiteboard, although they shouldn't be used at the same time.

* Actionbar / overflow menu is not recalculated on device orientation. Due to ankidroid handling orientation changes.

* Setting screen does not immediately update on "reset to default"

timrae commented on AnkiDroid/src/main/res/values/10-preferences.xml:
```diff
@@ -141,6 +141,15 @@
     <string name="convert_fen_text_summ">Draw chessboard from Forsyth–Edwards Notation. The notation must be enclosed in [fen][/fen] tags.</string>
     <string name="enable_api_title">Enable AnkiDroid API</string>
     <string name="enable_api_summary">Let other apps connect to AnkiDroid and read / write data without your intervention</string>
+    <string name="custom_buttons">Customize buttons</string>
```
I feel some of the Strings should be slightly changed as follows:
"Customize buttons" -> "Toolbar buttons"
"... reviewer action bar" -> "... reviewer toolbar"
"If room" -> "Show if room"

timrae commented on AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:
```diff
@@ -1664,6 +1675,25 @@ protected SharedPreferences restorePreferences() {
+        mCustomButtons.put(R.id.action_replay, Integer.parseInt(preferences.getString("customButtonReplay", "1")));
+        mCustomButtons.put(R.id.action_clear_whiteboard, Integer.parseInt(preferences.getString("customButtonClearWhiteboard", "2")));
+        mCustomButtons.put(R.id.action_hide_whiteboard, Integer.parseInt(preferences.getString("customButtonShowHideWhiteboard", "2")));
+        mCustomButtons.put(R.id.action_enable_whiteboard, Integer.parseInt(preferences.getString("customButtonEnableWhiteboard", "0")));
```
Enable whiteboard should never go in the toolbar, it should be Menu only or Disabled as it's not intended to be used often and it would be confusing having it shown at the same time as show/hide whiteboard as you previously mentioned

timrae commented on AnkiDroid/src/main/java/com/ichi2/anki/Reviewer.java:
```diff
@@ -351,12 +419,29 @@ public void fillFlashcard() {
+        if(note.cards().size() > 1) {
+            ArrayList<Card> cards = note.cards();
+            for(Card card : cards) {
+                if(card.getQueue() != Card.QUEUE_SUSP && card.getQueue() != Card.QUEUE_USER_BRD && card.getQueue() != Card.QUEUE_SCHED_BRD && card.getId() != mCurrentCard.getId())
```
Please always use braces for your `if` statements; we don't use the contracted form in this project. Also, please try to ensure that you do not exceed the 140 character lime limit guideline. It's OK to exceed the 140 character limit in XML files and is occasionally permissible in Java files, but in general you split long statements across multiple lines, or use an extra variable.

timrae commented on AnkiDroid/src/main/java/com/ichi2/anki/Reviewer.java:
```diff
@@ -202,11 +236,58 @@ public boolean onOptionsItemSelected(MenuItem item) {
     @SuppressLint("NewApi")
     @Override
     public boolean onCreateOptionsMenu(Menu menu) {
+        mMenu = menu;
```
If possible, I don't think you should store a reference to the menu. The reference can easily become invalid and I experienced several crashes due to null pointer exceptions on this variable during testing.

timrae commented on AnkiDroid/src/main/java/com/ichi2/anki/Reviewer.java:
```diff
@@ -202,11 +236,58 @@ public boolean onOptionsItemSelected(MenuItem item) {
+                menu.findItem(itemId).setVisible(false);
+            }
+        }
+        // Workaround for submenu items "If room".
```
I feel like this code is overly complicated and a bit hard to read. More informative comments would help, but how about we just get rid of the "Hide / delete" submenu in favor and replace it with the three new actions (Bury / Suspend / Delete note)? I believe a lot of new users were having trouble discovering the bury feature anyway.

timrae commented on AnkiDroid/src/main/res/xml/preferences_custom_buttons.xml:
```diff
@@ -0,0 +1,113 @@
+            android:key="customButtonClearWhiteboard"
+            android:title="@string/clear_whiteboard" />
+        <ListPreference
+            android:defaultValue="2"
```
Having four items set to "Always show" by default is too much on small screens; we need to make one of them default to "If room". How about we make "Hide whiteboard" use "If room" by default? This is how it used to be in a previous version of AnkiDroid before I added the workaround which is getting removed by this PR

timrae commented on AnkiDroid/src/main/res/xml/preferences_custom_buttons.xml:
```diff
@@ -0,0 +1,113 @@
+            android:defaultValue="2"
+            android:entries="@array/custom_button_labels"
+            android:entryValues="@array/custom_button_values"
+            android:key="customButtonMarkCard"
```
Should be `customButtonMarkNote`

timrae commented on AnkiDroid/src/main/java/com/ichi2/anki/Reviewer.java:
```diff
@@ -351,12 +419,29 @@ public void fillFlashcard() {
         // show timer, if activated in the deck's preferences
         initTimer();
         super.displayCardQuestion();
+        if(mMenu.findItem(R.id.action_bury_actionbar_only).isVisible() || mMenu.findItem(R.id.action_suspend_actionbar_only).isVisible())
```
Updating the menu every time a new card is shown may significantly lower performance, and this is the place where I was experiencing crashes due to null pointer exceptions. If possible, whether or not we show a submenu should be calculated when the button is pressed. I believe this should be quite easy to achieve by using an action provider instead of an embedded menu.

timrae commented on AnkiDroid/src/main/java/com/ichi2/anki/Reviewer.java:
```diff
@@ -132,6 +136,18 @@ public boolean onOptionsItemSelected(MenuItem item) {
+                    mMenu.findItem(R.id.action_bury_actionbar_only).getSubMenu().setGroupVisible(R.id.group_menu_bury_actionbar_only, false);
+                    DeckTask.launchDeckTask(DeckTask.TASK_TYPE_DISMISS_NOTE, mDismissCardHandler, new DeckTask.TaskData(mCurrentCard, 4));
+                }
+                else {
```
Please contract to `} else {` to be consistent with the rest of the code base

timrae commented:
Thanks a lot for all your work on this! I've added a bunch of comments inline. Some of them are simple, but two of them would involve quite a lot of work. Please let me know if you don't have time to make some of the bigger changes, as maybe myself or someone else could help out.

arooooon commented:
Your suggestions are fine, I will add them. Shouldn't be too much work.

About the menu reference, android docs says:
> You can safely hold on to menu (and any items created from it), making modifications to it as desired, until the next time onCreateOptionsMenu() is called. 
So it seems it should have worked. When onCreateOptionsMenu is called again the reference would just update? Somehow displayCardQuestion must be called when mMenu is null. Or could menu be null during onOptionsItemSelected?

One of the reasons I initially added it to the displayCardQuestion was because I thought it would be good to have the bury and suspend icons have an indication when they will take immediate action and when they will show a submenu. It is always nice to have a predictable UI. 

> More informative comments would help, but how about we just get rid of the "Hide / delete" submenu and replace it with the three new actions (Bury / Suspend / Delete note)?
You mean to have it gone by default and with the same functionality as in the toolbar? Then I really think we should have some kind of indication as well. The icons could be underlined and the text could have an ellipse "bury..." when there is a choice to be made. Or a toast would be great.

timrae commented:
> You mean to have it gone by default and the buttons having the same functionality as in the toolbar? 

Umm it's kind of unclear what you mean by that. What I mean is completely remove the "Hide / Delete" menu from the xml code. It's functionality would entirely be replaced with the three (Bury/Suspend/Delete Note) items, which can each be set to (Always/If Room/Menu only/Disabled)

timrae commented:
> Then I really think we should have some kind of indication as well. The icons could be underlined and the text could have an ellipse "bury..." when there is a choice to be made. Or a toast would be great.

I'm just throwing ideas around, but how about showing a snackbar when there are other cards available to be suspended / buried saying something like "Card suspended" together with a "Suspend note" action? Snackbars can be swiped off the screen so they're pretty non-invasive.

timrae commented:
> About the menu reference, android docs says [...] so it seems it should have worked.

Hmmm OK I'm not sure what the root cause of the NPE is, but if we can make it so that the action bar isn't changed every time the card is updated, then keeping the reference might not be necessary anymore. Otherwise we'll need to debug the NPE.

hssm commented:
>Snackbars can be swiped off the screen so they're pretty non-invasive.

In the reviewer, snackbars would cover the answer buttons.

timrae commented:
> In the reviewer, snackbars would cover the answer buttons.

Ah true, good point! We could create a new behavior to animate them above the snackbar like the floating add button in the deck picker does, but I think that's not so elegant.

I guess it wouldn't be too bad to update the menu and icon every time the question is shown, since this work is only done when the user has explicitly configured either the suspend or delete action to be promoted to the toolbar. Latency shouldn't be affected for users that haven't done this.

> The icons could be underlined and the text could have an ellipse "bury..." when there is a choice to be made.

Rather than underline, how about a drop-down arrow in the Action Provider style like in the images below? Rather than an ellipsis how about just switching the text between e.g. "Bury" (more than one card available) and "Bury card" (only one card available). For performance and readability reasons, I think this switching of the text should be done in `onPrepareOptionsMenu()` which is called just before the menu is shown.

![image](https://cloud.githubusercontent.com/assets/2818274/12806519/25c12510-cb4a-11e5-858c-ed846ef5b8f3.png)

arooooon commented:
Sorry for being so slow to answer all the time ;)

I really liked the snackbar idea, but with the answer field it will look kinda ugly indeed. But the drop-down arrow is a good indicator.

> I guess it wouldn't be too bad (in terms of latency) to update at least the menu icon every time the question is shown (we actually already do this for mark note), **so long as the work is only done when the user has explicitly configured either the suspend or delete action to be promoted to the toolbar.**

Hmm so you don't think we should change the menu text onPrepareOptionsMenu()? We don't know exactly how much lag that will result in, but if it is too much then we can just remove that or optimize it somehow. Few people have more than maybe 3 cards in their notes?


Yes, let's get rid of the submenu.

timrae commented:
Changing the text in onPrepareOptionsMenu() is all good, I'm just saying that we should only change the icon to include the drop-down arrow when the action is actually indeed visible in the toolbar.

Storing a reference may be necessary, that's OK I guess but let's figure out why the exception was occurring before and make sure that doesn't happen.

arooooon commented:
Just wanna say I am still working on this, but I have been really tired and lazy these past few weeks :)

timrae commented:
Good to hear, it's almost there!
On 25 Feb 2016 09:25, "arooooon" <notifications@github.com> wrote:

> Just wanna say I am still working on this, but I have been really tired
> and lazy these past few weeks :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ankidroid/Anki-Android/pull/4058#issuecomment-188532578>
> .
>

arooooon commented:
About the custom actionprovider... To create two different classes for suspend and bury is very uneccessary ofc, but how should I define in the xml which of these types I want? Some custom attribute perhaps? Is that even possible with custom actionproviders?

timrae commented:
I think it's probably fine to use two different classes actually... They'll both need separate implementations of `hasSubMenu()`, `onPrepareSubMenu()`, and `onMenuItemClick()` so the amount of duplicate code should be quite minimal.

arooooon commented:
Now that I have looked into it I think I am going to skip the action provider because I don't see at all how that would help or make the code cleaner.

Any ideas? Was it mainly meant for hiding and showing the submenu? Because as I see it you still would have to change the text and icon in Reviewer class.

timrae commented:
The idea was that with an action provider you shouldn't need to convert between an ordinary action and a submenu every time a new card is shown, you should be able to just override the hasSubmenu() method which is only called when the provider is tapped.

You're welcome to skip it for now though if you like.

timrae commented:
Any progress on this? :-D 

We're starting to discuss moving to beta soon and we do a feature freeze on beta release that would prevent this from being included in 2.6 if it were to miss the deadline.

timrae commented:
I happen to have some free time on my hands the next few days, so please let me take care of this. I'll open a new PR.

timrae commented:
Please check #4226

arooooon commented:
Ye sorry, I guess my time's up. I'm excited test your code later on.
Den 22 apr 2016 15:23 skrev "Tim Rae" <notifications@github.com>:

> Please check #4226 <https://github.com/ankidroid/Anki-Android/pull/4226>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/ankidroid/Anki-Android/pull/4058#issuecomment-213427002>
>
